README
README
ssterm - Version 1.6 - 02/11/2013
Vanya A. Sergeev - vsergeev at gmail
================================================================================
Table of Contents

1. ABOUT ssterm
2. INSTALLING
3. USAGE
4. LICENSE
5. USING ssterm
6. Example Usage
================================================================================

1. ABOUT ssterm
================================================================================

ssterm is a console-based simple serial port terminal. It features painless
serial port configuration, no dependencies outside of a standard Python 2
installation, is contained in a single file, and provides a variety of useful
input/output features:
	- plain ASCII representation
	- split hexadecimal/ASCII representation
	- pure hexadecimal representation with optional newline interpretation
	- color coding bytes in all three ASCII, split, and hexadecimal modes
	- raw terminal with control character forwarding
	- transmit newline character remapping (e.g. LF -> CRLF)
	- receive newline character remapping (e.g. CR -> LF)
	- local character echo

ssterm was written in Python 2, and should work on most *nix platforms. It
works with both CPython and pypy implementations of Python. Feel free to send
any bugs, ideas, or suggestions to vsergeev at gmail dot com.

2. INSTALLING
================================================================================

Install ssterm to /usr/local/bin (requires root):
# make install

Remove ssterm from /usr/local/bin (requires root):
# make uninstall

Install ssterm to an alternate prefix, such as /usr:
# make install PREFIX=/usr

3. USAGE
================================================================================

Usage: ssterm [options] <serial port device>

ssterm - simple serial-port terminal
Written by Vanya A. Sergeev - <vsergeev@gmail.com>.

 Serial Port Options:
  -b, --baudrate <rate>         Specify baudrate
  -d, --databits <number>       Specify number of data bits [5,6,7,8]
  -p, --parity <type>           Specify parity [none, odd, even]
  -t, --stopbits <number>       Specify number of stop bits [1,2]
  -f, --flow-control <type>     Specify flow-control [none, rtscts, xonxoff]

 Formatting Options:
  -s, --split                   Split hexadecimal/ASCII mode

  --split-full			Split hexadecimal/ASCII mode with full lines
                                 (good for piping)

  -x, --hex                     Pure hexadecimal mode
  --hex-nl                      Print newlines in pure hexadecimal mode

  -c, --color <list>            Specify comma-delimited list of characters in
                                 ASCII or hex to color code: A,$,0x0d,0x0a,...

  --tx-nl <substitution>        Specify the transmit newline substitution
                                  [raw, none, cr, lf, crlf]
  --rx-nl <match>               Specify the receive newline match
                                  [raw, cr, lf, crlf, crorlf]

  -e, --echo                    Turn on local character echo

  -h, --help                    Display this usage/help
  -v, --version                 Display the program's version


Quit Escape Character:          Ctrl-]

Color Code Sequence (fg/bg):
 Black/Red, Black/Green, Black/Yellow, White/Blue, White/Magenta,
 Black/Cyan, Black/White

Default Options:
 baudrate: 9600 | databits: 8 | parity: none | stopbits: 1 | flow ctrl: none
 tx newline: raw | rx newline: raw | local echo: off
 split mode: off | hex mode: off   | color code: off


4. LICENSE
================================================================================

ssterm is released under the GNU General Public License Version 3.
    You should have received a copy of the GNU General Public License
    along with this program; see the file "COPYING".  If not, see
    <http://www.gnu.org/licenses/>.

5. USING ssterm
================================================================================

By default, ssterm will open the specified serial port with 9600 baudrate, 8
data bits, no parity, 1 stop bit, and no flow control. These settings can be
adjusted with the -b / --baudrate, -d / --databits, -p / --parity, -t /
--stopbits, and -f / --flow-control options (see usage below).

ssterm opens the input terminal in raw mode, so all characters and control
signals are passed through transparently to the serial port. This enables
interacting with terminal interfaces over the serial port that interpret
control characters, such as a *nix termios on an embedded system. The only
exception is Ctrl-], which serves as ssterm's quit escape character.

ssterm can display output in split hexadecimal/ASCII mode, with the -s or
--split option. This mode will format data in the split mode as it is received,
updating the hexadecimal and ASCII columns by clearing and re-drawing the line
until it is full width.  For a split hexadecimal/ASCII representation suitable
for piping to another program or file, --split-full should be used, to limit
printing to full column lines.

ssterm can color code particular characters, specified by a comma-delimited
list with the -c or --color option. The list can contain ASCII entries (e.g.
a,_,A,...), as well as hexadecimal entries (e.g. 0x0a,0xff,...). There is
currently a maximum of seven distinct colors, assigned sequentially to the
specified characters in the following order (foreground/background): Black/Red,
Black/Green, Black/Yellow, White/Blue, White/Magenta, Black/Cyan, Black/White.

ssterm can also display output in pure hexadecimal mode, with the -x or --hex
option, and interpret newlines in this mode with the --hex-nl option.

Plain ASCII representation mode is default.

The newline characters can be remapped upon transmission and remapped upon
reception. These mappings are specified with the --tx-nl and --rx-nl options.
These options were created to accommodate working with the serial port on LF
based systems, but the functionality should extend to CR and CRLF based systems
as well. The substitution and matching can be disabled by selecting 'raw' mode
for both --tx-nl and --rx-nl options.

The following receive newline character remappings are available:
	* raw - no remapping of received newline characters
	* cr - map received CR to platform newline
	* lf - map received LF to platform newline
	* crlf - map received CRLF to platform newline
	* crorlf - map any received CR or LF to platform newline

The following transmit newline character remappings are available:
	* raw - no remapping of transmit newline characters
	* none - omit newline characters on transmit
	* cr - map platform newline to CR on transmit
	* lf - map platform newline to LF on transmit
	* crlf - map platform newline to CRLF on transmit

The default receive newline character mapping (--rx-nl option) is raw.
The default transmit newline character mapping (--rx-nl option) is raw.

Other options include local character echo, which is off by default.

6. Example Usage
================================================================================

Typical usage with defaults (9600 8N1, no flow control):
$ ssterm /dev/ttyUSB0

Logging some data:
$ ssterm /dev/ttyUSB0 > nmea_data.txt

Viewing and logging some data, with the help of tee:
$ ssterm /dev/ttyUSB0 | tee nmea_data.txt

Specific serial port settings (115200 baudrate, odd parity, 2 stop bits), and
the default 8 data bits:
$ ssterm -b 115200 -p odd -t 2 /dev/ttyUSB0

CRLF for transmitted newlines, but LF for received newlines:
$ ssterm --tx-nl crlf --rx-nl lf /dev/ttyUSB0

Split representation mode:
$ ssterm -s /dev/ttyUSB0
4c 6f 72 65 6d 20 69 70  73 75 6d 20 64 6f 6c 6f  |Lorem ipsum dolo|
72 20 73 69 74 20 61 6d  65 74 2c 20 63 6f 6e 73  |r sit amet, cons|
65 63 74 65 74 75 72 20  61 64 69 70 69 73 69 63  |ectetur adipisic|
69 6e 67 20 65 6c 69 74  2c 20 73 65 64 20 64 6f  |ing elit, sed do|
20 65 69 75 73 6d 6f 64  20 74 65 6d 70 6f 72 20  | eiusmod tempor |
69 6e 63 69 64 69 64 75  6e 74 20 75 74 20 6c 61  |incididunt ut la|
62 6f 72 65 20 65 74 20  64 6f 6c 6f 72 65 20 6d  |bore et dolore m|
61 67 6e 61 20 61 6c 69  71 75 61 2e 0a 55 74 20  |agna aliqua..Ut |
65 6e 69 6d 20 61 64 20  6d 69 6e 69 6d 20 76 65  |enim ad minim ve|
6e 69 61 6d 2c 20 71 75  69 73 20 6e 6f 73 74 72  |niam, quis nostr|
75 64 20 65 78 65 72 63  69 74 61 74 69 6f 6e 20  |ud exercitation |
75 6c 6c 61 6d 63 6f 20  6c 61 62 6f 72 69 73 20  |ullamco laboris |
6e 69 73 69 20 75 74 20  61 6c 69 71 75 69 70 20  |nisi ut aliquip |
65 78 20 65 61 20 63 6f  6d 6d 6f 64 6f 20 63 6f  |ex ea commodo co|
6e 73 65 71 75 61 74 2e  0a 44 75 69 73 20 61 75  |nsequat..Duis au|
74 65 20 69 72 75 72 65  20 64 6f 6c 6f 72 20 69  |te irure dolor i|
6e 20 72 65 70 72 65 68  65 6e 64 65 72 69 74 20  |n reprehenderit |
69 6e 20 76 6f 6c 75 70  74 61 74 65 20 76 65 6c  |in voluptate vel|
69 74 20 65 73 73 65 20  63 69 6c 6c 75 6d 20 64  |it esse cillum d|
6f 6c 6f 72 65 20 65 75  20 66 75 67 69 61 74 20  |olore eu fugiat |
6e 75 6c 6c 61 20 70 61  72 69 61 74 75 72 2e 0a  |nulla pariatur..|
45 78 63 65 70 74 65 75  72 20 73 69 6e 74 20 6f  |Excepteur sint o|
63 63 61 65 63 61 74 20  63 75 70 69 64 61 74 61  |ccaecat cupidata|
74 20 6e 6f 6e 20 70 72  6f 69 64 65 6e 74 2c 20  |t non proident, |
73 75 6e 74 20 69 6e 20  63 75 6c 70 61 20 71 75  |sunt in culpa qu|
69 20 6f 66 66 69 63 69  61 20 64 65 73 65 72 75  |i officia deseru|
6e 74 20 6d 6f 6c 6c 69  74 20 61 6e 69 6d 20 69  |nt mollit anim i|

ChangeLog
Release 1.6 - 02/11/2013
 - Switched to gnu_getopt to allow command-line options after serial port
   device argument.
 - Refactored codebase.

Release 1.5 - 04/04/2012
 - Fixed escape character bug. Switched escape character from Ctrl-[ to the
   more unique Ctrl-], which does not serve as the escape code for many other
   special keys and cause ssterm to quit on them like Ctrl-[ did. Thanks to
   zzazang for discovering this bug.
 - Modified formatting of split hexadecimal/ASCII representation mode to
   conform to "hexdump -C" canonical split output.

Release 1.4 - 03/28/2012
 - Added support for split hexadecimal/ASCII representation mode.
 - Added controlling terminal on after program quit.
 - Fixed non-blocking read bug.

Release 1.3 - 03/19/2012
 - Added support for color coding characters / bytes.
 - Rewrote ssterm in Python 2 for ease of future extensibility and
   maintenance.
 - Upgraded license from GPLv2 to GPLv3.

Release 1.2 - 02/04/2011
 - Added mutexes for safer handling of shared variables across threads.

Release 1.1 - 11/22/2009
 - CRITICAL FIX: Serial port was not being opened in non-blocking mode,
   preventing ssterm from working in some situations. This has been
   fixed.

Release 1.0 - 10/26/2009
 - Initial release.


ssterm.py
#!/usr/bin/python2

# ssterm - simple serial-port terminal
# Version 1.6 - February 2013
# Written by Vanya A. Sergeev - <vsergeev@gmail.com>
#
# Copyright (C) 2007-2013 Vanya A. Sergeev
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#

import sys
import os
import termios
import getopt
import select
import re
import string

# Default TTY and Formatting Options
TTY_Options = {'baudrate': 9600, 'databits': 8, 'stopbits': 1, 'parity': "none", 'flowcontrol': "none"}
Format_Options = {'splitmode': False, 'splitfullmode': False, 'hexmode': False, 'txnl': "raw", 'rxnl': "raw", 'hexnl': False, 'echo': False}
Color_Chars = {}
Console_Newline = os.linesep

# Number of columns in hex mode
Hexmode_Columns = 16

# ssterm Quit Escape Character, Ctrl-] = 0x1D
Quit_Escape_Character = 0x1D

# Default color codes: Black/Red, Black/Green, Black/Yellow, White/Blue,
#  White/Magenta, Black/Cyan, Black/White
Color_Codes = ["\x1b[1;30;41m", "\x1b[1;30;42m", "\x1b[1;30;43m", "\x1b[1;37;44m", "\x1b[1;37;45m", "\x1b[1;30;46m", "\x1b[1;30;47m"]
Color_Code_Reset = "\x1b[0m"

# Newline Matches / Substitutions
RX_Newline_Match = {'raw': None, 'cr': "\r", 'crlf': "\r\n", 'lf': "\n", 'crorlf': "\r|\n"}
TX_Newline_Sub = {'raw': None, 'cr': "\r", 'crlf': "\r\n", 'lf': "\n", 'none': ""}

# Read buffer size
READ_BUFF_SIZE = 4096

###########################################################################
### Serial Helper Functions
###########################################################################

def serial_open(devpath, tty_options):
    # Open the tty device
    try:
        tty_fd = os.open(devpath, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK);
    except OSError, err:
        sys.stderr.write("Error: opening serial port: %s\n" % str(err))
        return -1

    # Get the current tty options
    # Format: [iflag, oflag, cflag, lflag, ispeed, ospeed, cc]
    try:
        tty_attr = termios.tcgetattr(tty_fd)
    except termios.TermiosError, err:
        sys.stderr.write("Error: getting serial port options: %s\n" % str(err))
        return -1

    # Look up the termios baudrate and set it in the attributes structure
    termios_baudrates = {50: termios.B50, 75: termios.B75, 110: termios.B110, 134: termios.B134, 150: termios.B150, 200: termios.B200, 300: termios.B300, 600: termios.B600, 1200: termios.B1200, 1800: termios.B1800, 2400: termios.B2400, 4800: termios.B4800, 9600: termios.B9600, 19200: termios.B19200, 38400: termios.B38400, 57600: termios.B57600, 115200: termios.B115200, 230400: termios.B230400}
    if (not tty_options['baudrate'] in termios_baudrates):
        sys.stderr.write("Error: Invalid tty baudrate!\n")
        return -1
    tty_attr[4] = termios_baudrates[tty_options['baudrate']]
    tty_attr[5] = termios_baudrates[tty_options['baudrate']]

    # Reset attributes structure cflag -- tty_attribute[cflag]
    tty_attr[2] = 0

    # Look up and set the appropriate cflag bits in termios_options for a
    # given option, print error message and return -1 for an invalid option
    def termios_cflag_map_and_set(termios_options, option, errmsg):
        if (not option in termios_options):
            sys.stderr.write(errmsg)
            return -1
        tty_attr[2] |= termios_options[option]
        return 0

    # Look up the termios data bits and set it in the attributes structure
    termios_databits = {5: termios.CS5, 6: termios.CS6, 7: termios.CS7, 8: termios.CS8}
    if (termios_cflag_map_and_set(termios_databits, tty_options['databits'], "Error: Invalid tty databits!\n") < 0):
        return -1

    # Look up the termios parity and set it in the attributes structure
    termios_parity = {"none": 0, "even": termios.PARENB, "odd": termios.PARENB | termios.PARODD}
    if (termios_cflag_map_and_set(termios_parity, tty_options['parity'], "Error: Invalid tty parity!\n") < 0):
        return -1

    # Look up the termios stop bits and set it in the attributes structure
    termios_stopbits = {1: 0, 2: termios.CSTOPB}
    if (termios_cflag_map_and_set(termios_stopbits, tty_options['stopbits'], "Error: Invalid tty stop bits!\n") < 0):
        return -1

    # Look up the termios flow control and set it in the attributes structure
    termios_flowcontrol = {"none": 0, "rtscts": termios.CRTSCTS, "xonxoff": 0}
    if (termios_cflag_map_and_set(termios_flowcontrol, tty_options['flowcontrol'], "Error: Invalid tty flow control!\n") < 0):
        return -1

    # Enable the receiver
    tty_attr[2] |= (termios.CREAD | termios.CLOCAL);

    # Turn off signals generated for special characters, turn off canonical
    # mode so we can have raw input -- tty_attr[lflag]
    tty_attr[3] = 0

    # Turn off POSIX defined output processing and character mapping/delays
    # so we can have raw output -- tty_attr[oflag]
    tty_attr[1] = 0

    # Ignore break characters -- tty_attr[iflag]
    tty_attr[0] = termios.IGNBRK
    # Enable parity checking if we are using parity -- tty_attr[iflag]
    if (tty_options['parity'] != "none"):
        tty_attr[0] |= (termios.INPCK | termios.ISTRIP)
    # Enable XON/XOFF if we are using software flow control
    if (tty_options['flowcontrol'] == "xonxoff"):
        tty_attr[0] |= (termios.IXON | termios.IXOFF | termios.IXANY)

    # Set the new tty attributes
    try:
        termios.tcsetattr(tty_fd, termios.TCSANOW, tty_attr)
    except termios.TermiosError, err:
        sys.stderr.write("Error: setting serial port options: %s\n" % str(err))
        return -1

    # Return the tty_fd
    return tty_fd

def serial_close(fd):
    try:
        os.close(fd)
        return 0
    except:
        return -1

###########################################################################
### TTY Helper Functions
###########################################################################

def stdin_raw_open(echo):
    stdin_fd = sys.stdin.fileno()

    # Get the current stdin tty options
    # Format: [iflag, oflag, cflag, lflag, ispeed, ospeed, cc]
    try:
        stdin_attr = termios.tcgetattr(stdin_fd)
    except termios.TermiosError, err:
        sys.stderr.write("Error: getting stdin tty options: %s\n" % str(err))
        return -1

    # Disable canonical input, so we can send characters without a line
    # feed, disable signal interpretation, and disable echo
    # -- stdin_attr[cflag]
    stdin_attr[3] &= ~(termios.ICANON | termios.ECHO | termios.ECHOE | termios.ISIG)
    # Turn off XON/XOFF interpretation so they pass through to the serial
    # port -- stdin_attr[iflag]
    stdin_attr[0] &= ~(termios.IXON | termios.IXOFF | termios.IXANY)
    # Enable echo if needed
    if (echo):
        stdin_attr[3] |= termios.ECHO

    # Set the new stdin tty attributes
    try:
        termios.tcsetattr(stdin_fd, termios.TCSANOW, stdin_attr)
    except termios.TermiosError, err:
        sys.stderr.write("Error: setting stdin tty options: %s\n" % str(err))
        return -1

    return stdin_fd

def stdout_raw_open():
    # Re-open stdout in unbuffered mode
    sys.stdout = os.fdopen(sys.stdout.fileno(), 'w', 0)

    return sys.stdout.fileno()

def stdin_reset():
    stdin_fd = sys.stdin.fileno()

    # Reset stdin terminal for canonical input, echo, and signals

    # Get the current stdin tty options
    # Format: [iflag, oflag, cflag, lflag, ispeed, ospeed, cc]
    try:
        stdin_attr = termios.tcgetattr(stdin_fd)
    except termios.TermiosError, err:
        sys.stderr.write("Error: getting stdin tty options: %s\n" % str(err))
        return -1

    # Enable canonical input, echo, signals -- stdin_attr[cflag]
    stdin_attr[3] |= (termios.ICANON | termios.ECHO | termios.ECHOE | termios.ISIG)

    # Set the new stdin tty attributes
    try:
        termios.tcsetattr(stdin_fd, termios.TCSANOW, stdin_attr)
    except termios.TermiosError, err:
        sys.stderr.write("Error: setting stdin tty options: %s\n" % str(err))
        return -1

    return 0

###########################################################################
### fd read and write
###########################################################################

def fd_read(fd, n):
    try:
        return (0, os.read(fd, n))
    except OSError, err:
        # Check if non-blocking read returned 0
        if (err.errno == 11): return (0, None)
        else: return (-1, str(err))

def fd_write(fd, data):
    try:
        return (os.write(fd, data), None)
    except OSError, err:
        return (-1, str(err))

###########################################################################
### Receive and transmit newline substitutions
###########################################################################

# State used by format_rx_sub
rxnl_match_save = ''

def format_tx_sub(data, sub):
    # FIXME: This assumes a single character platform newline.
    data = map(lambda x: sub if x == Console_Newline else x, list(data))
    data = ''.join(data)

    return data

def format_rx_sub(data, match):
    global rxnl_match_save

    # Append left-over newline character match from before
    data = rxnl_match_save + data
    rxnl_match_save = ''

    # Substitute newline matches with console newline
    data = re.sub(match, Console_Newline, data)

    # If the last character is a part of a match, save it for later
    if data[-1] == match[0][0]:
        rxnl_match_save = data[-1]
        data = data[0:-1]

    return data

###########################################################################
### Formatted Print
###########################################################################

# State used by format_print_*mode
stdout_cursor_x = 0
stdout_split_bytes = []

def format_print_normal(stdout_fd, data):
    # Apply Color coding if necessary
    if len(Color_Chars) > 0:
        # Unfortunately, for generality, we can't do a global
        # regex substitution on data with the color-coded
        # version, since we could have potentially selected
        # color code characters that are present in the ANSI
        # color escape sequences. So we operate on the data
        # a char at time here.
        for x in list(data):
            # Color code this character if it's in our color chars dictionary
            if ord(x) in Color_Chars:
                fd_write(stdout_fd, Color_Codes[Color_Chars[ord(x)]] + x + Color_Code_Reset)
            else:
                fd_write(stdout_fd, x)
    else:
        fd_write(stdout_fd, data)

def format_print_hexmode(stdout_fd, data):
    global stdout_cursor_x

    for x in list(data):
        # Color code this character if it's in our color chars dictionary
        if len(Color_Chars) > 0 and ord(x) in Color_Chars:
            fd_write(stdout_fd, Color_Codes[Color_Chars[ord(x)]] + ("%02x" % ord(x)) + Color_Code_Reset)
        else:
            fd_write(stdout_fd, "%02x" % ord(x))
        stdout_cursor_x += 1

        # Pretty print into two columns
        if stdout_cursor_x == Hexmode_Columns/2:
            fd_write(stdout_fd, "  ")
        elif stdout_cursor_x == Hexmode_Columns:
            fd_write(stdout_fd, "\n")
            stdout_cursor_x = 0
        else:
            fd_write(stdout_fd, " ")

        # Insert a newline if we encounter one and we're
        # interpreting them in hex mode
        # FIXME: This assumes a single character platform newline.
        if x == Console_Newline and Format_Options['hexnl']:
            fd_write(stdout_fd, Console_Newline)
            stdout_cursor_x = 0

def format_print_splitmode(stdout_fd, data, partial_print=True):
    global stdout_split_bytes

    # Erase partially completed strings with \r
    if partial_print:
        if len(stdout_split_bytes) > 0:
            fd_write(stdout_fd, "\r")

    def split_print(byte_list):
        # split_print() expects:
        # 1 <= len(byte_list) <= Hexmode_Columns

        # Print the hexadecimal representation
        for i in range(len(byte_list)):
            # Color code this character if it's in our color chars dictionary
            if len(Color_Chars) > 0 and ord(byte_list[i]) in Color_Chars:
                fd_write(stdout_fd, Color_Codes[Color_Chars[ord(byte_list[i])]] + ("%02x" % ord(byte_list[i])) + Color_Code_Reset)
            else:
                fd_write(stdout_fd, "%02x" % ord(byte_list[i]))

            # Pretty print into two columns
            if (i+1) == Hexmode_Columns/2:
                fd_write(stdout_fd, "  ")
            else:
                fd_write(stdout_fd, " ")

        # Fill up the rest of the hexadecimal representation
        # with blank space
        if len(byte_list) < Hexmode_Columns/2:
            # Account for the pretty print column separator
            fd_write(stdout_fd, " " + " "*(3*(Hexmode_Columns-len(byte_list))))
        elif len(byte_list) < Hexmode_Columns:
            fd_write(stdout_fd, " "*(3*(Hexmode_Columns-len(byte_list))))

        # Print the ASCII representation
        fd_write(stdout_fd, " |")
        for i in range(len(byte_list)):
            # Use the character if it's an ASCII printable
            # character, otherwise use a dot
            if (byte_list[i] in string.letters+string.digits+string.punctuation+' '):
                c = byte_list[i]
            else:
                c = "."
            # Color code this character if it's in our
            # color chars dictionary
            if len(Color_Chars) > 0 and ord(byte_list[i]) in Color_Chars:
                fd_write(stdout_fd, Color_Codes[Color_Chars[ord(byte_list[i])]] + c + Color_Code_Reset)
            else:
                fd_write(stdout_fd, c)
        fd_write(stdout_fd, "|")

    for x in list(data):
        # Add to our split byte window
        stdout_split_bytes.append(x)
        # If we get a full column window, print it out with a
        # newline
        if (len(stdout_split_bytes) == Hexmode_Columns):
            split_print(stdout_split_bytes)
            fd_write(stdout_fd, Console_Newline)
            stdout_split_bytes = []

    # Print partially completed strings
    if partial_print:
        # Print out any bytes left in our window
        if len(stdout_split_bytes) > 0:
            split_print(stdout_split_bytes)

###########################################################################
### Read/Write Loop
###########################################################################

def read_write_loop(serial_fd, stdin_fd, stdout_fd):
    # Look up our TX newline sub and RX newline match
    txnl_sub = TX_Newline_Sub[Format_Options['txnl']]
    rxnl_match = RX_Newline_Match[Format_Options['rxnl']]

    # Select our format print function
    if Format_Options['hexmode']:
        format_print = format_print_hexmode
    elif Format_Options['splitmode']:
        format_print = format_print_splitmode
    elif Format_Options['splitfullmode']:
        format_print = lambda s, d: format_print_splitmode(s, d, False)
    else:
        format_print = format_print_normal

    # Select between serial port and stdin file descriptors
    read_fds = [serial_fd, stdin_fd]

    while True:
        ready_read_fds, ready_write_fds, ready_excep_fds = select.select(read_fds, [], [])

        if stdin_fd in ready_read_fds:
            # Read a buffer from stdin
            retval, buff = fd_read(stdin_fd, READ_BUFF_SIZE)
            if retval < 0:
                sys.stderr.write("Error: reading stdin: %s\n" % buff)
                break
            if buff and len(buff) > 0:
                # Perform transmit newline subsitutions
                if txnl_sub != None:
                    buff = format_tx_sub(buff, txnl_sub)

                # If we detect the escape character, then quit
                if chr(Quit_Escape_Character) in buff:
                    break

                # Write the buffer to the serial port
                retval, err = fd_write(serial_fd, buff)
                if retval < 0:
                    sys.stderr.write("Error: writing to serial port: %s\n" % err)
                    break

        if serial_fd in ready_read_fds:
            # Read a buffer from the serial port
            retval, buff = fd_read(serial_fd, READ_BUFF_SIZE)
            if retval < 0:
                sys.stderr.write("Error: reading serial port: %s\n" % buff)
                break

            # Format and print the buffer to the console
            if buff and len(buff) > 0:
                # Perform receive newline substitutions
                if rxnl_match != None:
                    buff = format_rx_sub(buff, rxnl_match)
                # Print to stdout
                format_print(stdout_fd, buff)

###########################################################################
### Command-Line Options Parsing
###########################################################################

def print_usage():
    print "Usage: %s [options] <serial port device>\n" % sys.argv[0]
    print "\
ssterm - simple serial-port terminal\n\
Written by Vanya A. Sergeev - <vsergeev@gmail.com>.\n\
\n\
 Serial Port Options:\n\
  -b, --baudrate <rate>         Specify baudrate\n\
  -d, --databits <number>       Specify number of data bits [5,6,7,8]\n\
  -p, --parity <type>           Specify parity [none, odd, even]\n\
  -t, --stopbits <number>       Specify number of stop bits [1,2]\n\
  -f, --flow-control <type>     Specify flow-control [none, rtscts, xonxoff]\n\
\n\
 Formatting Options:\n\
  -s, --split                   Split hexadecimal/ASCII mode\n\
\n\
  --split-full                  Split hexadecimal/ASCII mode with full lines\n\
                                  (better for piping than --split)\n\
\n\
  -x, --hex                     Pure hexadecimal mode\n\
  --hex-nl                      Print newlines while in pure hexadecimal mode\n\
\n\
  -c, --color <list>            Specify comma-delimited list of characters in\n\
                                  ASCII or hex to color code: A,$,0x0d,0x0a,...\n\
\n\
  --tx-nl <substitution>        Specify transmit newline substitution\n\
                                  [raw, none, cr, lf, crlf]\n\
  --rx-nl <match>               Specify receive newline match\n\
                                  [raw, cr, lf, crlf, crorlf]\n\
\n\
  -e, --echo                    Turn on local character echo\n\
\n\
  -h, --help                    Display this usage/help\n\
  -v, --version                 Display the program's version\n\n"
    print "\
Quit Escape Character:          Ctrl-]\n\
\n\
Color Code Sequence (fg/bg):\n\
 Black/Red, Black/Green, Black/Yellow, White/Blue, White/Magenta,\n\
 Black/Cyan, Black/White\n\
\n\
Default Options:\n\
 baudrate: 9600 | databits: 8 | parity: none | stopbits: 1 | flow ctrl: none\n\
 tx newline: raw | rx newline: raw | local echo: off\n\
 split mode: off | hex mode: off   | color code: off\n"

def print_version():
    print "ssterm version 1.6 - 02/11/2013"

# Parse options
try:
    options, args = getopt.gnu_getopt(sys.argv[1:], "b:d:p:t:f:esxhvc:", ["baudrate=", "databits=", "parity=", "stopbits=", "flowcontrol=", "tx-nl=", "rx-nl=", "echo", "split", "split-full", "hex", "hex-nl", "color-nl", "help", "version", "color="])
except getopt.GetoptError, err:
    print str(err), "\n"
    print_usage()
    sys.exit(-1)

# Update options containers
for opt_c, opt_arg in options:
    if opt_c in ("-b", "--baudrate"):
        try:
            TTY_Options['baudrate'] = int(opt_arg, 10)
        except:
            sys.stderr.write("Error: Invalid tty baudrate!\n")
            sys.exit(-1)

    elif opt_c in ("-d", "--databits"):
        try:
            TTY_Options['databits'] = int(opt_arg, 10)
        except:
            sys.stderr.write("Error: Invalid tty data bits!\n")
            sys.exit(-1)

    elif opt_c in ("-p", "--parity"):
        TTY_Options['parity'] = opt_arg

    elif opt_c in ("-t", "--stopbits"):
        try:
            TTY_Options['stopbits'] = int(opt_arg, 10)
        except:
            sys.stderr.write("Error: Invalid tty stop bits!\n")
            sys.exit(-1)

    elif opt_c in ("-f", "--flowcontrol"):
        TTY_Options['flowcontrol'] = opt_arg

    elif opt_c in ("-e", "--echo"):
        Format_Options['echo'] = True

    elif opt_c in ("-s", "--split"):
        Format_Options['splitmode'] = True

    elif opt_c in ("--split-full"):
        Format_Options['splitfullmode'] = True

    elif opt_c in ("-x", "--hex"):
        Format_Options['hexmode'] = True

    elif opt_c in ("-c", "--color"):
        opt_arg = filter(lambda x: len(x) >= 1, opt_arg.split(","))
        if len(opt_arg) > len(Color_Codes):
            sys.stderr.write("Error: Maximum color code characters (%d) exceeded!\n" % len(Color_Codes))
            sys.exit(-1)
        # Parse ASCII and hex encoded characters into our Color_Chars dictionary
        for c in opt_arg:
            # ASCII character
            if len(c) == 1:
                Color_Chars[ord(c)] = len(Color_Chars)
            # Hexadecimal number
            elif len(c) > 2 and c[0:2] == "0x":
                try:
                    c_int = int(c, 16)
                except:
                    sys.stderr.write("Error: Unknown color code character: %s\n" % c)
                    sys.exit(-1)
                Color_Chars[c_int] = len(Color_Chars)
            # Unknown
            else:
                sys.stderr.write("Error: Unknown color code character: %s\n" % c)
                sys.exit(-1)

    elif opt_c == "--tx-nl":
        Format_Options['txnl'] = opt_arg
        if (not Format_Options['txnl'] in TX_Newline_Sub):
            sys.stderr.write("Error: Invalid tx newline type!\n")
            print_usage()
            sys.exit(-1)

    elif opt_c == "--rx-nl":
        Format_Options['rxnl'] = opt_arg
        if (not Format_Options['rxnl'] in RX_Newline_Match):
            sys.stderr.write("Error: Invalid rx newline type!\n")
            print_usage()
            sys.exit(-1)

    elif opt_c == "--hex-nl":
        Format_Options['hexnl'] = True

    elif opt_c in ("-h", "--help"):
        print_usage()
        sys.exit(0)

    elif opt_c in ("-v", "--version"):
        print_version()
        sys.exit(0)


# Make sure a serial port device is specified
if len(args) < 1:
    print_usage()
    sys.exit(-1)

# Open the serial port with our options
serial_fd = serial_open(args[0], TTY_Options)
if (serial_fd < 0):
    sys.exit(-1)

# Open stdin in raw mode
stdin_fd = stdin_raw_open(Format_Options['echo'])
if (stdin_fd < 0):
    sys.exit(-1)

# Open stdout in raw mode
stdout_fd = stdout_raw_open()
if (stdout_fd < 0):
    sys.exit(-1)

read_write_loop(serial_fd, stdin_fd, stdout_fd)
fd_write(stdout_fd, "\n")

# Reset our console to buffered mode
stdin_reset()

# Close the serial port
serial_close(serial_fd)

